test/lib/jdk/test/lib/hotspot/ir_framework/IRMatcher.java

```diff
@@ -0,0 +1,562 @@
+     */
+    private void setupTestMethods(String irEncoding) {
+        Map<String, Integer[]> irRulesMap = parseIREncoding(irEncoding);
+        for (Method m : testClass.getDeclaredMethods()) {
+            method = m;
+            IR[] irAnnos =  m.getAnnotationsByType(IR.class);
+            if (irAnnos.length > 0) {
+                // Validation of legal @IR attributes and placement of the annotation was already done in Test VM.
+                Integer[] ids = irRulesMap.get(m.getName());
+                TestFramework.check(ids != null, "Should find method name in validIrRulesMap for " + m);
+                TestFramework.check(ids.length > 0, "Did not find any rule indices for " + m);
+                TestFramework.check(ids[ids.length - 1] < irAnnos.length, "Invalid IR rule index found in validIrRulesMap for " + m);
+                if (ids[0] != IREncodingPrinter.NO_RULE_APPLIED) {
+                    // If -1, than there was no matching IR rule for the given conditions.
+                    compilations.put(m.getName(), new IRMethod(m, ids, irAnnos));
+                }
+            }
+        }
+    }
+
+    /**
+     * Read the IR encoding emitted by the test VM to decide if an @IR rule must be checked for a method.
+     */
+    private Map<String, Integer[]>  parseIREncoding(String irEncoding) {
+        Map<String, Integer[]> irRulesMap = new HashMap<>();
+        String patternString = "(?<=" + IREncodingPrinter.START + "\\R)[\\s\\S]*(?=" + IREncodingPrinter.END + ")";
+        Pattern pattern = Pattern.compile(patternString);
+        Matcher matcher = pattern.matcher(irEncoding);
+        TestFramework.check(matcher.find(), "Did not find IR encoding");
+        String[] lines = matcher.group(0).split("\\R");
+
+        // Skip first line containing information about the format only
+        for (int i = 1; i < lines.length; i++) {
+            String line = lines[i].trim();
+            String[] splitComma = line.split(",");
+            if (splitComma.length < 2) {
+                throw new TestFrameworkException("Invalid IR match rule encoding");
+            }
+            String testName = splitComma[0];
+            Integer[] irRulesIdx = new Integer[splitComma.length - 1];
+            for (int j = 1; j < splitComma.length; j++) {
+                irRulesIdx[j - 1] = Integer.valueOf(splitComma[j]);
+            }
+            irRulesMap.put(testName, irRulesIdx);
+        }
+        return irRulesMap;
+    }
+
+    /**
+     * Parse the hotspot_pid*.log file from the test VM. Read the PrintIdeal and PrintOptoAssembly entries for all
+     * methods of the test class that need to be IR matched (according to IR encoding).
+     */
+    private void parseHotspotPidFile() {
+        Map<Integer, String> compileIdMap = new HashMap<>();
+        try (BufferedReader br = new BufferedReader(new FileReader(new File(System.getProperty("user.dir") + File.separator + hotspotPidFileName)))) {
```

Comment:
more idiomatic/modern version would be
```suggestion
try (BufferedReader br = Files.newBufferedReader(Paths.get(System.getProperty("user.dir"), hotspotPidFileName))) {
```

I actually not sure if you really need `$user.dir`, won't hotspot_pid file get generated in the current dir?

Reply:
`Paths.get(hotspotPidFileName)` returns you a `Path` to `hotspotPidFileName` in cwd.